packages/http-client-csharp/generator/docs/pagination-design.md

```diff
@@ -0,0 +1,343 @@
+#### TypeSpec
+```ts
+@list op listPets(@query @pageSize perPage?: int32): {
+  @pageItems pets: Pet[];
+  @nextLink next?: url;
+};
+```
+#### Client APIs
+
+```c#
+public class PetClient 
+{
+    // protocol methods
+    public virtual CollectionResult GetPets(int? perPage = default(int?), RequestOptions options = default(RequestOptions));
+    public virtual AsyncCollectionResult GetPets(int? perPage = default(int?), RequestOptions options = default(RequestOptions));
+    
+    // convenience methods
+    public virtual CollectionResult<Pet> GetPets(int? perPage = default(int?), CancellationToken cancellationToken = default(CancellationToken));
+    public virtual AsyncCollectionResult<Pet> GetPetsAsync(int? perPage = default(int?), CancellationToken cancellationToken = default(CancellationToken));
+}
+
+internal class GetPetsCollectionResult : CollectionResult, CollectionResult<Pet>, AsyncCollectionResult, AsyncCollectionResult<T>
+{
+    public GetPetsCollectionResult(ClientPipeline pipeline, Uri endpoint, int? perPage, CancellationToken cancellationToken);
+    public override ContinuationToken? GetContinuationToken(ClientResult page);
+    public override IEnumerable<ClientResult> GetRawPages();
+    public override async IAsyncEnumerable<ClientResult> GetRawPagesAsync();
+    protected override IEnumerable<Pet> GetValuesFromPage(ClientResult page);
+    protected override AsyncEnumerable<Pet> GetValuesFromPageAsync(ClientResult page);
+}
+
+internal class GetPetsCollectionResultPage : IJsonModel<InnerModel>
+{
+    public IList<Pet> Pets { get; }
+    public Uri Next { get; }
+    public static explicit operator GetPetsCollectionResultPage(ClientResult result);
+}
+```
+
+#### Example Client Usage
+
+```c#
+PetClient client = new(...);
+
+// sync convenience usage
+foreach (Pet p in client.GetPets()) { ... }
+
+// async convenience usage
+await foreach (Pet p in client.GetPetsAsync(perPage: 20)) { ... }
+
+// sync protocol usage
+foreach (ClientResult page in client.GetPets(perPage: 10, options: requestOptions)) { ... }
+
+// async protocol usage
+await foreach (ClientResult page in client.GetPetsAsync(options: requestOptions)) { ... }
+```
+
+#### Example Client Implementation
+
+```c#
+public class PetClient 
+{
+    // protocol methods
+    public virtual CollectionResult GetPets(int? perPage = default(int?), RequestOptions options = default(RequestOptions));
+    public virtual AsyncCollectionResult GetPets(int? perPage = default(int?), RequestOptions options = default(RequestOptions));
+     
+    // convenience methods
+    public virtual CollectionResult<Pet> GetPets(int? perPage = default(int?), CancellationToken cancellationToken = default(CancellationToken));
+    public virtual AsyncCollectionResult<Pet> GetPetsAsync(int? perPage = default(int?), CancellationToken cancellationToken = default(CancellationToken));
+}
+
+internal class GetPetsCollectionResult : CollectionResult<Pet>, AsyncCollectionResult<T>
```

Comment:
We would need separate implementations for sync and async since CollectionResult is an abstract class and not an interface.